File: MCP服务器/content-extraction/AmplifyCo_botbrowser_fbdc244_typescript/python/botbrowser/cleaner.py
"""HTML cleaning pipeline — strips bloat from web pages."""

from __future__ import annotations

from bs4 import BeautifulSoup, Comment, Tag

REMOVE_TAGS = {
    "script", "style", "noscript", "iframe", "object", "embed",
    "applet", "svg", "canvas", "video", "audio", "source", "track",
    "map", "area",
}

REMOVE_SELECTORS = [
    "nav",
    "[role='navigation']",
    "[role='banner']",
    "[role='complementary']",
    "[aria-hidden='true']",
    ".ad", ".ads", ".advertisement", ".adsbygoogle",
    ".sidebar", ".side-bar",
    ".cookie-banner", ".cookie-notice", ".cookie-consent",
    ".popup", ".modal", ".overlay",
    ".social-share", ".share-buttons", ".social-links",
    ".comments", ".comment-section", "#comments",
    ".newsletter", ".subscribe",
    ".breadcrumb", ".breadcrumbs",
    ".pagination",
    ".related-posts", ".related-articles",
    ".widget",
    "[data-ad]",
    "[data-tracking]",
]

STRIP_ATTR_PREFIXES = ("data-", "aria-", "on")

SELF_CLOSING = {"img", "br", "hr", "input", "meta", "link"}


def clean_html(html: str) -> str:
    """Remove non-content elements and unnecessary attributes from HTML."""
    soup = BeautifulSoup(html, "html.parser")

    # Remove HTML comments
    for comment in soup.find_all(string=lambda t: isinstance(t, Comment)):
        comment.extract()

    # Remove unwanted tags entirely
    for tag_name in REMOVE_TAGS:
        for tag in soup.find_all(tag_name):
            tag.decompose()

    # Remove non-content elements by selector
    for selector in REMOVE_SELECTORS:
        try:
            for el in soup.select(selector):
                el.decompose()
        except Exception:
            pass

    # Remove hidden elements
    for el in soup.find_all(True):
        if not isinstance(el, Tag):
            continue
        if el.has_attr("hidden"):
            el.decompose()
            continue
        style = el.get("style", "")
        if isinstance(style, str) and (
            "display:none" in style.replace(" ", "")
            or "visibility:hidden" in style.replace(" ", "")
        ):
            el.decompose()

    # Strip unnecessary attributes
    for el in soup.find_all(True):
        if not isinstance(el, Tag):
            continue
        attrs_to_remove = []
        for attr_name in list(el.attrs.keys()):
            lower = attr_name.lower()
            if lower in ("style", "class", "id", "role", "tabindex", "draggable", "contenteditable"):
                attrs_to_remove.append(attr_name)
            elif any(lower.startswith(p) for p in STRIP_ATTR_PREFIXES):
                attrs_to_remove.append(attr_name)
        for attr_name in attrs_to_remove:
            del el[attr_name]

    # Remove empty elements
    for el in soup.find_all(True):
        if not isinstance(el, Tag):
            continue
        if (
            el.name not in SELF_CLOSING
            and not el.get_text(strip=True)
            and not el.find("img")
        ):
            el.decompose()

    return str(soup)

File: reference-mcp-servers/content-extraction/AmplifyCo_botbrowser_fbdc244_typescript/js/packages/core/src/cleaner.ts
import { JSDOM } from "jsdom";

/** Tags that are always removed entirely (including content) */
const REMOVE_TAGS = [
  "script",
  "style",
  "noscript",
  "iframe",
  "object",
  "embed",
  "applet",
  "svg",
  "canvas",
  "video",
  "audio",
  "source",
  "track",
  "map",
  "area",
];

/** Selectors for common non-content elements */
const REMOVE_SELECTORS = [
  "nav",
  "header:not(article header)",
  "footer:not(article footer)",
  "[role='navigation']",
  "[role='banner']",
  "[role='complementary']",
  "[aria-hidden='true']",
  ".ad, .ads, .advertisement, .adsbygoogle",
  ".sidebar, .side-bar",
  ".cookie-banner, .cookie-notice, .cookie-consent",
  ".popup, .modal, .overlay",
  ".social-share, .share-buttons, .social-links",
  ".comments, .comment-section, #comments",
  ".newsletter, .subscribe",
  ".breadcrumb, .breadcrumbs",
  ".pagination",
  ".related-posts, .related-articles",
  ".widget",
  "[data-ad]",
  "[data-tracking]",
];

/** Attributes to strip from all remaining elements */
const STRIP_ATTRIBUTES = [
  "style",
  "class",
  "id",
  "data-*",
  "onclick",
  "onload",
  "onerror",
  "onmouseover",
  "onmouseout",
  "onfocus",
  "onblur",
  "aria-*",
  "role",
  "tabindex",
  "draggable",
  "contenteditable",
];

export function cleanHtml(html: string, url: string): string {
  const dom = new JSDOM(html, { url });
  const document = dom.window.document;

  // Remove unwanted tags entirely
  for (const tag of REMOVE_TAGS) {
    const elements = document.querySelectorAll(tag);
    elements.forEach((el) => el.remove());
  }

  // Remove non-content elements by selector
  for (const selector of REMOVE_SELECTORS) {
    try {
      const elements = document.querySelectorAll(selector);
      elements.forEach((el) => el.remove());
    } catch {
      // Some selectors might fail on malformed HTML — skip
    }
  }

  // Remove hidden elements
  const allElements = document.querySelectorAll("*");
  allElements.forEach((el) => {
    const htmlEl = el as HTMLElement;
    if (
      htmlEl.getAttribute("hidden") !== null ||
      htmlEl.getAttribute("style")?.includes("display:none") ||
      htmlEl.getAttribute("style")?.includes("display: none") ||
      htmlEl.getAttribute("style")?.includes("visibility:hidden") ||
      htmlEl.getAttribute("style")?.includes("visibility: hidden")
    ) {
      el.remove();
    }
  });

  // Strip unnecessary attributes from remaining elements
  const remaining = document.querySelectorAll("*");
  remaining.forEach((el) => {
    const attrs = Array.from(el.attributes);
    for (const attr of attrs) {
      const name = attr.name.toLowerCase();
      if (
        STRIP_ATTRIBUTES.includes(name) ||
        name.startsWith("data-") ||
        name.startsWith("aria-") ||
        name.startsWith("on")
      ) {
        el.removeAttribute(attr.name);
      }
    }
  });

  // Remove empty elements (except self-closing ones like img, br, hr)
  const selfClosing = new Set(["img", "br", "hr", "input", "meta", "link"]);
  const emptyCheck = document.querySelectorAll("*");
  emptyCheck.forEach((el) => {
    if (
      !selfClosing.has(el.tagName.toLowerCase()) &&
      !el.textContent?.trim() &&
      !el.querySelector("img")
    ) {
      el.remove();
    }
  });

  return document.body?.innerHTML || "";
}
